feature(Grade management)

lomamech opened the pull request:
- Implement client side for Grade Management
- Implementation of test E2E for Grade Management

reference #2292

lomamech committed:
Add grade Payroll arborescence

lomamech committed:
feature(Grade management)
- Update translate key

lomamech committed:
feature(Grade Management)
- Implement client side for Grade Management

lomamech committed:
test(Grade Management)
- Implementation of test E2E for Grade Management

jniles commented on client/src/i18n/en/grade.json:
```diff
@@ -0,0 +1,26 @@
+    "DELETED" : "Grade deleted with success",
+    "DELETE" : "Delete",
+    "GRADES_LIST" : "Grades list",
+    "DESCRIPTION" : "This module allows you to List, Create, Edit and Delete a Grade",
```
It doesn't look like you use the DESCRIPTION.  It could be removed.

jniles commented on client/src/modules/grades/grades.js:
```diff
@@ -0,0 +1,104 @@
+      { field : 'basic_salary',
+        displayName : 'FORM.LABELS.BASIC_SALARY',
+        headerCellFilter : 'translate',
+        cellFilter : 'currency:'.concat(Session.enterprise.currency_id),
```
I wonder if we should allow payroll to be in multiple currencies?  I can imagine that a hospital might want to pay some employees in USD and some in FC.

lomamech replied:
In bhima 1.X the basic salary was always defined with the currency of the enterprise, but when the agents were paid the currency could be specified individually for each employee

jniles replied:
Okay, that sounds like a good default.  When we actually install somewhere, we'll wee what our clients wish.

jniles commented on client/src/modules/grades/templates/action.tmpl.html:
```diff
@@ -0,0 +1,22 @@
+    <span class="caret"></span>
+  </a>
+
+  <ul data-action="{{rowRenderIndex}}"  class="dropdown-menu-right" uib-dropdown-menu>
```
You should use `bh-dropdown-menu-auto-dropup` here to make sure that if there are many grades, the dropdown will continue to be visible.

jniles commented on client/src/modules/grades/grades.routes.js:
```diff
@@ -0,0 +1,42 @@
+    })
+
+    .state('grades.edit', {
+      url : '/edit',
```
Most `edit` routes include the ID of the record they are editing.  Could you do that here?

jniles commented on client/src/modules/grades/grades.js:
```diff
@@ -0,0 +1,104 @@
+
+  // update an existing grade
+  function editGrade(gradeObject) {
+    $state.go('grades.edit', { grade : gradeObject });
```
It would be better to simply pass the ID and load the grade at from the modal.  This will make it so that if a user reloads the page on the modal, the latest data is downloaded instead having a blank edit state.

jniles commented on client/src/modules/grades/templates/action.tmpl.html:
```diff
@@ -0,0 +1,22 @@
+    <span class="caret"></span>
+  </a>
+
+  <ul data-action="{{rowRenderIndex}}"  class="dropdown-menu-right" uib-dropdown-menu>
```
Do you ever use `rowRenderIndex`? If not, it should be removed.

lomamech replied:
It's used in E2E test

jniles commented on test/end-to-end/grades/grades.spec.js:
```diff
@@ -0,0 +1,48 @@
+const chai = require('chai');
+
+/** loading User pages **/
+const UserPage = require('../user/user.page.js');
```
It doesn't look like you use these `User` imports.  Can you remove them please?

jniles commented on test/end-to-end/grades/grades.spec.js:
```diff
@@ -0,0 +1,48 @@
+
+/** configuring helpers**/
+helpers.configure(chai);
+const expect = chai.expect;
```
These two lines are unnecessary.  Go ahead and remove them.

jniles commented on client/src/modules/grades/modals/grade.modal.html:
```diff
@@ -0,0 +1,51 @@
+    </button>
+
+    <bh-loading-button loading-state="GradeForm.$loading">
+      <i class="fa fa-ok"></i> <span translate>FORM.BUTTONS.SUBMIT</span>
```
Most of our form buttons don't have icons on them.  Let's keep to that standard by removing these icons.

jniles commented on client/src/modules/grades/modals/grade.modal.html:
```diff
@@ -0,0 +1,51 @@
+  <ol class="headercrumb">
+    <li ng-if="GradeModalCtrl.isCreating" class="title">
+      <span translate>GRADE.ADD_GRADE</span>
+      <label class="badge badge-warning" translate>FORM.LABELS.CREATE</label>
```
I think most `CREATE` labels in our application use `badge-success`.

jniles reviewed (commented):
Looks pretty good!  I've made a few clean-up comments.  Let me know what you think!

lomamech committed:
perf(Grade Management)
- Sanitaze translate key
- Update the updating process for Grade Management
- Sanitaze E2E test

jniles commented on client/src/modules/grades/grades.js:
```diff
@@ -0,0 +1,104 @@
+
+  // update an existing grade
+  function editGrade(grade) {
+    $state.go('grades.edit', { uuid : grade.uuid });
```
Nice :+1:

jniles commented on client/src/modules/grades/grades.routes.js:
```diff
@@ -0,0 +1,42 @@
+    })
+
+    .state('grades.edit', {
+      url : '/:uuid/edit',
```
Nice :+1:

jniles commented on client/src/modules/grades/grades.routes.js:
```diff
@@ -0,0 +1,42 @@
+    .state('grades.edit', {
+      url : '/:uuid/edit',
+      params : {
+        grade : { value : null },
```
I think this is probably supposed to be `uuid` now.  But if it works without it, then it might not be necessary.

jniles reviewed (approved):
These changes look good to me.  Great work 👍 

bors r+
